core/chains/evm/client/tx_simulator.go

```diff
@@ -0,0 +1,134 @@
+	"github.com/smartcontractkit/chainlink/v2/common/config"
+
+	commonclient "github.com/smartcontractkit/chainlink/v2/common/client"
+)
+
+const ErrOutOfCounters = "not enough counters to continue the execution"
+
+type simulatorClient interface {
+	CallContext(ctx context.Context, result interface{}, method string, args ...interface{}) error
+}
+
+// ZK chains can return an out-of-counters error
+// This method allows a caller to determine if a tx would fail due to OOC error by simulating the transaction
+// Used as an entry point for custom simulation across different chains
+func SimulateTransaction(ctx context.Context, client simulatorClient, lggr logger.SugaredLogger, chainType config.ChainType, msg ethereum.CallMsg) error {
+	var err error
+	switch chainType {
+	case config.ChainZkEvm:
+		err = simulateTransactionZkEvm(ctx, client, lggr, msg)
+	default:
+		err = simulateTransactionDefault(ctx, client, msg)
+	}
+	// ClassifySendError will not have the proper fields for logging within the method due to the empty Transaction passed
+	code := ClassifySendError(err, lggr, &types.Transaction{}, msg.From, chainType.IsL2())
+	// Only return error if ZK OOC error is identified
+	if code == commonclient.OutOfCounters {
+		return errors.New(ErrOutOfCounters)
+	}
+	return nil
+}
+
+// eth_estimateGas returns out-of-counters (OOC) error if the transaction would result in an overflow
+func simulateTransactionDefault(ctx context.Context, client simulatorClient, msg ethereum.CallMsg) error {
+	var result hexutil.Big
+	errCall := client.CallContext(ctx, &result, "eth_estimateGas", toCallArg(msg), "pending")
+	jsonErr, _ := ExtractRPCError(errCall)
+	if jsonErr != nil && len(jsonErr.Message) > 0 {
+		return errors.New(jsonErr.Message)
+	}
+	return nil
+}
+
+type zkEvmEstimateCountResponse struct {
+	CountersUsed struct {
+		GasUsed              string
+		UsedKeccakHashes     string
+		UsedPoseidonHashes   string
+		UsedPoseidonPaddings string
+		UsedMemAligns        string
+		UsedArithmetics      string
+		UsedBinaries         string
+		UsedSteps            string
+		UsedSHA256Hashes     string
+	}
+	CountersLimit struct {
+		MaxGasUsed          string
+		MaxKeccakHashes     string
+		MaxPoseidonHashes   string
+		MaxPoseidonPaddings string
+		MaxMemAligns        string
+		MaxArithmetics      string
+		MaxBinaries         string
+		MaxSteps            string
+		MaxSHA256Hashes     string
+	}
+	OocError string
+}
+
+// zkEVM implemented a custom zkevm_estimateCounters method to detect if a transaction would result in an out-of-counters (OOC) error
+func simulateTransactionZkEvm(ctx context.Context, client simulatorClient, lggr logger.SugaredLogger, msg ethereum.CallMsg) error {
+	var result zkEvmEstimateCountResponse
+	err := client.CallContext(ctx, &result, "zkevm_estimateCounters", toCallArg(msg), "pending")
+	if err != nil {
+		return fmt.Errorf("failed to simulate tx: %w", err)
+	}
+	if detectZkEvmCounterOverflow(result) && len(result.OocError) > 0 {
```

Comment:
The `&&` condition here seems redundant, might be better to err on the side of caution and make it a `||`?
The error message might be changed in the future and not match the regex and this way we could still detect it.

Reply:
I realized I may have misunderstood the original [comment](https://github.com/smartcontractkit/chainlink/pull/12503#discussion_r1536223363) that prompted me to add `detectZkEvmCounterOverflow`. I don't think we want to completely break reliance on the error message. We want to pass it through our `ClassifySendError` [method](https://github.com/smartcontractkit/chainlink/pull/12503/files#diff-ed85d15901e7cfbfdfa625b140d903c8ec415cf09c24373330e695b874b3053aR36). To your point if they change the message and our regex no longer matches it, you're right we'd stop detecting the overflow but we have some work in progress to allow regex's to be configurable through toml which would allow us to quickly address this issue in the future. If we solely relied on the used vs max comparison from the response, we'd maybe run the risk of a new counter getting added that we don't account for. Don't know how likely that'd be but the only path forward there would be a code change limiting how quickly we could address it.

Reply:
Since we're passing this through our `ClassifySendError` method, I think it's ok if we just pass the error message through if one exists without checking further. To avoid the risk of counter fields changing, I'm feeling inclined to remove `detectZkEvmCounterOverflow`.